### PR TITLE
Added features to do Sanic more pluggable (example included)

### PR DESCRIPTION
**Pull request by cr0hn**
This PR try to add some features that already exists in other frameworks (like aiohttp) and that are useful for the plugin development:
- add te "app" property into each request, with te Sanic instances 
- add the property "handler" in each request, that contains the view handed where the request was originated 
- new method to add a middleware: add_middelware

The features do not affect to the performance (I was tested) and the could be very useful try to grow the framework and create plugins like:

Example of a A cache system
---------------------------

```python

import asyncio
import uvloop

from aiocache import RedisCache, MemcachedCache, SimpleMemoryCache
from aiocache.serializers import JsonSerializer

#
# Decorator
#
class cache(object):
    def __init__(self, expires: int = 3600, unless: bool = False, use_post_data = False):
        self.expires = expires
        self.unless = unless
        self.use_post_data = use_post_data
    
    def __call__(self, f):
        f.cache_enable = True
        f.cache_expires = self.expires
        f.cache_unless = self.unless
        f.cache_user_post_data = self.use_post_data
        
        return f

#
# Middleware
#
async def cache_middleware(request):
    _handler = request.handler
    try:
        if getattr(_handler, "cache_enable", False):
            #
            # Cache is disabled?
            #
            if getattr(_handler, "cache_unless", False) is True:
                return await _handler(request)
        
            cache_backend = request.app.config.cache
        
            #
            # Make key
            #
            key = "{method}#{path}#{post_data}".format(
                method=request.method,
                path=request.url,
                post_data=request.body if getattr(_handler, "cache_user_post_data") else "")
        
            cached_response = await cache_backend.get(key)
            if cached_response:
                return HTTPResponse(**cached_response)
        
            #
            # Generate cache
            #
            original_response = await _handler(request)
        
            data = dict(status=original_response.status,
                        content_type=original_response.content_type,
                        headers=original_response.headers,
                        body=original_response.body)
        
            expires = getattr(_handler, "cache_expires", 300)
        
            await cache_backend.set(key, data, ttl=expires)
        
            return original_response
    except Exception as e:
        return text(str(e))

    # Not cached
    return await _handler(request)


#
# Cache configuration
#
def setup_cache(app: Sanic,
                cache_type: str = "memory",
                loop=None):
    _cache_backend = None
    if cache_type.lower() == "memory":
        _cache_backend = SimpleMemoryCache(serializer=JsonSerializer())
    
    elif cache_type.lower() == "redis":
        _cache_backend = RedisCache(serializer=JsonSerializer(), loop=loop)
    
    elif cache_type.lower() == "memcached":
        _cache_backend = MemcachedCache(serializer=JsonSerializer(), loop=loop)
    
    else:
        raise HTTPResponse("Invalid cache type selected")
    
    app.config.cache = _cache_backend
    app.add_middleware(cache_middleware)

#
# Main
#   
asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())

loop = asyncio.get_event_loop()

app = Sanic()
    
@app.route("/")
@cache()
async def test(request):
    return json({"hello": "world"})


setup_cache(app, loop=loop, cache_type="memory")
# setup_cache(app, loop=loop, cache_type="redis")
    
app.run(host='127.0.0.1',
        port=8000,
        loop=loop)    
```

### CODE REVIEW

File: sanic/sanic.py [hunk 1]
```diff
@@ -168,6 +187,10 @@ def converted_response_type(self, response):
         :return: Nothing
         """
         try:
+
+            # Add view handler
+            request.handler = self.router.routes_static.get(request.url)
```
Comment (seemethere) [inline sanic/sanic.py]: Would this still work with a dynamic route as well?

**Comment by seemethere** [comment]
Also why add the app to the server? What's the point of it?

File: sanic/sanic.py (hunk 1 above)
Comment (cr0hn) [inline sanic/sanic.py]: I tried with some end-points with dynamic routes and it works fine.

**Comment by cr0hn** [comment]
The point to add the app to the server is really to add to the request object. The reasons are some:

- Access to the shared event loop object in views / middlewares.
- Access to configuration in views.

To access to the above information we need to has access to the `Sanic` instance object. But, probability we want to create a medium-big app using Sanic, and we need to use a scalable structure, like: http://exploreflask.com/en/latest/organizing.html

This organisation was made for Flask and has not in count some things, because the framework provide other methods to obtain this information:

1- Access to the app object in any place: `Flask` provide `current_app` object. But Sanic has not way to do that. If we can't access to the app object we can't access to the global configuration and plugins will have not do things like:

```python
from sanic import Sanic

app = Sanic()

def setup_db(app: Sanic,
             connection_string='postgresql://postgres:postgres@127.0.0.1/test',
             loop=None):
   # STORE DATABASE POOL HANDLER
    app.config.db_engine = loop.run_until_complete(create_engine(connection_string,
                                                                 minsize=10,
                                                                 maxsize=20,
                                                                 timeout=10,
                                                                 loop=loop))

@app.route("/q")
async def query(request):
        results = []
        
        async with request.app.config.db_engine.acquire() as conn:  # ACCESS IN ALL END-POINTS TO DB-POOL.
            async for row in conn.execute("SELECT 1;"):
                results.append(dict(id=row.id))
        
        return json(results)

app.run()
```

This way made each view / middleware self-contained, without the need of external shared objects.

2 - Some times we need to access to the event loop object in the views / middlewares. To do that also need to access to the `Sanic` instance. 

This is especially useful when we run Sanic with more than 1 worker. Sanic executes one instance in different processes. Each process has their own event loop. So, if we do something like that, it break:

```python

import asyncio
import uvloop

from sanic import Sanic

asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())
    
loop = asyncio.get_event_loop()
app = Sanic()
    
@app.route("/")
async def test(request):
    url = "https://api.github.com/repos/channelcat/sanic"

   # USE EVENT-LOOP GENERATED IN ANOTHER PROCESS -> BREAK THE APP
    async with aiohttp.ClientSession(loop=loop) as session:
        response = await fetch(session, url)
        return json(response)

app.run(host='127.0.0.1',
            port=8000,
            workers=4)
```

So we probably need to access to the event loop object in some view / middleware, so we need the loop reference used in the current instance:

```python
import aiohttp

from sanic import Sanic

app = Sanic()
    
@app.route("/")
async def test(request):
    url = "https://api.github.com/repos/channelcat/sanic"

   # SELF-CONTAINED VIEW. IT RUNS WITH MULTIPROCESS APPROACH
    async with aiohttp.ClientSession(loop=request.app.loop) as session:
        response = await fetch(session, url)
        return json(response)

app.run(host='127.0.0.1',
            port=8000,
            workers=4)
```

Applying this patch we can allow to do that, without performance impact. Also made more easy the development of extensions.

**Comment by gpip** [comment]
I miss the context local objects available in flask, some of the use cases would be handled by introducing a way to call a function after app.run starts and before requests are served (i.e. a middleware where request doesn't exist yet and is called only once).

But others I'm not sure yet how to handle other than working around them with uglier solutions, for instance I had to drop custom logging filters since those depended on the context local objects. Do you think it would be possible to add something equivalent to `flask.current_app`?

**Comment by cr0hn** [comment]
@seemethere What's about these new features? As time goes by there're more inconsistencies with the main trunk

**Comment by subyraman** [comment]
+1 to the ability to get the `current_app` via the loop. It will make accessing app configuration/routes much easier. I'd love to see this get in.

**Pull request closed without merging**